### PR TITLE
[Build] Add Github workflows

### PR DESCRIPTION
**Pull request by ktmud**
### CATEGORY

- [x] Build / Development Environment

### SUMMARY

Adding GitHub Actions per discussion in https://github.com/apache/incubator-superset/issues/9481. It is a pure addition to current build process for now. Nothing changes for Travis CI, except Cypress is upgraded from `3.6.1` to `4.3.0`.

We may evaluate whether to disable Travis CI once GitHub Actions become more stable. 

Jobs are split into 3 workflows to maximize parallelization:

- **superset-python.yml:** python lint and unit tests
- **superset-frontend.yml:** frontend lint and unit tests
- **superset-e2e.yml:** end to end tests with Cypress

A custom GitHub Action was created to generalize some of the building steps, but not all of them. There are duplicate code about cache across files, while could potentially be lifted into the GitHub Action like [the official Cypress GitHub action](
https://github.com/cypress-io/github-action/blob/d81aef503fcd11db601c9f79ca30d3353dcc46e9/index.js#L8).

#### Some thoughts

GitHub Actions is less powerful than some other CI/CD tools, e.g., CircleCI and GitLab CI/CD, but overall it was a pleasant experience to use. Two major nuances are: 1. limited parallelization support, cannot organize jobs as DAGs; 2. cannot automatically cancel current jobs on new pushes.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

image

### TEST PLAN

Make sure CI check pass.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

@john-bodley @dpgaspar @mistercrunch @craig-rueda

### CODE REVIEW

**Comment by ktmud** [comment]
See the builds in action: https://github.com/ktmud/incubator-superset/pull/25

File: .github/workflows/superset.yml [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    - superset/**/*.py
+    - tests/**
+    - setup.py
+    - requirements*.txt
+  # but also build on pull requests to any branch
+  # (the so-called feature branch)
+  pull_request:
+    paths:
+    - superset/**/*.py
+    - tests/**
```
Comment (dpgaspar) [inline .github/workflows/superset.yml:17]: Should this include `superset-frontend/**` ?

**Review by dpgaspar — COMMENTED** [review]
First pass

**Comment by craig-rueda** [comment]
Looks good - have you thought about Circle?

File: .github/workflows/superset.yml (hunk 1 above)
Comment (ktmud) [reply .github/workflows/superset.yml:17]: This file is purely for python code. Frontend code checks are done in other workflows. Maybe I this rename this file to avoid confusion.

**Comment by ktmud** [comment]
CircleCI was my first choice, but it seems there are some administration overheads involved in setting it up and no other Apache projects are using it AFAIK.

File: .github/workflows/superset-e2e.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        key: ${{ runner.os }}-pip-${{ hashFiles('./requirements*.txt') }}
+        restore-keys: |
+          ${{ runner.os }}-pip-
+    - name: Cache npm
+      uses: actions/cache@v1
+      with:
+        path: |
+          ~/.npm
+          superset-frontend/.terser-plugin-cache/
+        key: ${{ runner.os }}-npm-${{ hashFiles('superset-frontend/package-lock.json') }}
```
Comment (ktmud) [inline .github/workflows/superset-e2e.yml:49]: Cypress plugin will handle its own cache. So not need to add `cypress-base/package-lock.json` as hash key.

File: .github/workflows/superset-e2e.yml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      uses: cypress-io/github-action@v1
+      with:
+        working-directory: ./superset-frontend/cypress-base
+        # cannot use `superset run` because it starts the app somehow differently
+        # and would not respect the `SUPERSET_CONFIG` env
+        start: flask run -p 8081 --no-debugger
+        wait-on: http://localhost:8081
+        wait-on-timeout: 10
+        spec: cypress/integration/*/*
+        record: true
```
Comment (ktmud) [inline .github/workflows/superset-e2e.yml:67]: I'm enabling recording as it's free anyway. Check the recorded logs [here](https://dashboard.cypress.io/projects/dk2opw/).

File: superset-frontend/cypress-base/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+cypress/screenshots/
```
Comment (ktmud) [inline superset-frontend/cypress-base/.gitignore:1]: Local builds may generate this folder.

File: superset-frontend/cypress-base/cypress.json [hunk 5]
```diff
@@ -1,13 +1,14 @@
...
+  "defaultCommandTimeout": 5000,
+  "requestTimeout": 10000,
+  "ignoreTestFiles": [
+    "**/!(*.test.js)"
+  ],
   "video": false,
   "videoUploadOnPasses": false,
   "viewportWidth": 1280,
   "viewportHeight": 800,
-  "requestTimeout": 10000
```
Comment (ktmud) [inline superset-frontend/cypress-base/cypress.json:12]: The original config has duplicate `requestTimeout`.

File: superset-frontend/cypress-base/cypress.json [hunk 6]
```diff
@@ -1,13 +1,14 @@
 {
   "baseUrl": "http://localhost:8081",
   "chromeWebSecurity": false,
-  "defaultCommandTimeout": 20000,
-  "requestTimeout": 20000,
-  "ignoreTestFiles": ["**/!(*.test.js)"],
-  "projectId": "fbf96q",
+  "defaultCommandTimeout": 5000,
```
Comment (ktmud) [inline superset-frontend/cypress-base/cypress.json:4]: No need to actually wait for 20s for default commands (e.g., click triggers).

File: superset-frontend/cypress-base/cypress.json [hunk 7]
```diff
@@ -1,13 +1,14 @@
...
+  "requestTimeout": 10000,
+  "ignoreTestFiles": [
+    "**/!(*.test.js)"
+  ],
   "video": false,
   "videoUploadOnPasses": false,
   "viewportWidth": 1280,
   "viewportHeight": 800,
-  "requestTimeout": 10000
+  "projectId": "dk2opw"
```
Comment (ktmud) [inline superset-frontend/cypress-base/cypress.json:13]: Updated to a project under my personal account for testing. Will need to change back to [the official project ](https://dashboard.cypress.io/projects/fbf96q/runs) once merged.

Please [add a new record key](https://docs.cypress.io/guides/dashboard/projects.html#Set-up-a-project-to-record) to [GitHub Actions secrets](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets) under the name `CYPRESS_RECORD_KEY`.

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.js [hunk 8]
```diff
@@ -75,7 +75,7 @@ export default () =>
         .trigger('mousedown', { which: 1 })
         .trigger('dragstart', { dataTransfer })
         .trigger('drag', {});
-      cy.get('.grid-content .dragdroppable')
+      cy.get('.grid-content div.grid-row.background--transparent')
```
Comment (ktmud) [inline superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.js:78]: The upgraded Cypress will error out because it was dragged to a zero-height element. This fixes the error by dragging to the last visible row instead of the invisible one.

**Review by ktmud — COMMENTED** [review]
I think this is ready for another review. Everything finally works!

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.js [hunk 9]
```diff
@@ -23,37 +23,22 @@ export default () =>
...
         .contains('Edit dashboard')
         .click();
     });
 
     it('remove, and add chart flow', () => {
-      // wait box_plot data and find box plot
-      cy.wait('@boxplotRequest');
-      cy.get('.grid-container .box_plot').should('be.exist');
+      // wait for box plot to appear
+      cy.get('.grid-container .box_plot');
```
Comment (ktmud) [inline superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.js:34]: A simple `get` will wait for the element to appear. No need to wait for the additional request.

File: superset-frontend/cypress-base/cypress/integration/dashboard/load.js [hunk 10]
```diff
@@ -49,12 +49,15 @@ export default () =>
...
-          expect(responseBody).to.have.property('error', null);
-          cy.get(`#slice-container-${xhr.response.body.form_data.slice_id}`);
-        });
+        return Promise.all(
+          requests.map(async xhr => {
+            expect(xhr.status).to.eq(200);
+            const responseBody = await readResponseBlob(xhr.response.body);
+            expect(responseBody).to.have.property('error', null);
+            const sliceId = responseBody.form_data.slice_id;
+            cy.get(`#chart-id-${sliceId}`).should('be.visible');
```
Comment (ktmud) [inline superset-frontend/cypress-base/cypress/integration/dashboard/load.js:58]: Has to return a `Promise` for the test cases to be captured.

File: superset-frontend/cypress-base/cypress/integration/dashboard/save.js [hunk 11]
```diff
@@ -61,11 +60,7 @@ export default () =>
 
     it('should save/overwrite dashboard', () => {
       // should have box_plot chart
-      const formData = `{"slice_id":${boxplotChartId}}`;
-      const boxplotRequest = `/superset/explore_json/?form_data=${formData}&dashboard_id=${dashboardId}`;
-      cy.route('POST', boxplotRequest).as('boxplotRequest');
-      cy.wait('@boxplotRequest');
-      cy.get('.grid-container .box_plot').should('be.exist');
+      cy.get('.grid-container .box_plot', { timeout: 5000 }); // wait for 5 secs
```
Comment (ktmud) [inline superset-frontend/cypress-base/cypress/integration/dashboard/save.js:63]: Again, a simple `get` will wait. Also, `dashboardId` may be out of sync when test cases are running in parallelized mode.

File: superset-frontend/cypress-base/cypress/integration/dashboard/tabs.js [hunk 12]
```diff
@@ -190,11 +192,12 @@ export default () =>
       // trigger 1 new query
       cy.wait('@treemapRequest');
 
-      // no other requests occurred
+      // make sure query API not requested multiple times
       cy.on('fail', err => {
-        expect(err.message).to.include('Timed out retrying');
+        expect(err.message).to.include('timed out waiting');
```
Comment (ktmud) [inline superset-frontend/cypress-base/cypress/integration/dashboard/tabs.js:197]: The error message changed in Cypress v4.

**Review by ktmud — COMMENTED** [review]
Updated the Cypress test cases even more.

File: superset-frontend/cypress-base/cypress/integration/explore/visualizations/big_number.js [hunk 13]
```diff
@@ -56,8 +56,8 @@ export default () =>
 
     it('should work', () => {
       verify(BIG_NUMBER_FORM_DATA);
-      cy.get('.chart-container .header_line');
-      cy.get('.chart-container .subheader_line');
+      cy.get('.chart-container .header-line');
+      cy.get('.chart-container .subheader-line');
```
Comment (ktmud) [inline superset-frontend/cypress-base/cypress/integration/explore/visualizations/big_number.js:60]: `explore/visualizations/index.test.js` was probably unintentioally disabled by https://github.com/apache/incubator-superset/pull/6241 .

It has since stopped running for over a year now. A lot of these tests are now broken. It's easy to add them back, but we may want to do the fix in another PR. Keep the changes for BigNumber here just for reference.

**Comment by ktmud** [comment]
Woking on further improvements, closing for now to avoid excessive notification.

**Comment by willbarrett** [comment]
Hey @ktmud - let us know if there's anything we at Preset can do to support you on this work. We're really feeling the Travis pain and want to help get to a solution.
